doc/api/stream.md

```diff
@@ -443,9 +443,33 @@ first argument. To reliably detect write errors, add a listener for the
 The return value is `true` if the internal buffer does not exceed
 `highWaterMark` configured when the stream was created after admitting `chunk`.
 If `false` is returned, further attempts to write data to the stream should
-stop until the [`'drain'`][] event is emitted. However, the `false` return
-value is only advisory and the writable stream will unconditionally accept and
-buffer `chunk` even if it has not not been allowed to drain.
+stop until the [`'drain'`][] event is emitted.
+
+In case the `false` return value is ignored, the writable stream will
+unconditionally accept and buffer `chunk` even if it has not not been allowed
+to drain. However, Node has a limited amount of memory, and not
+listening to `'drain'` might cause the process to abort by an out of
+memory error, and other types of memory and garbage collection
+issues, such as high RSS.
+
+The `'drain'` event must be handled if the writable is a [Transform][],
+because the `Transform` streams are paused by default until
+they are piped or an `'data'` or `'readable'` event handler is added.
+So, ignoring when `write()` returns false will cause the data to be buffered
+until the stream is flowing, causing a memory leak.
+
+A possible way to handle the [`'drain'`][] event to respect
+backpressure is the following:
+
+```js
+function write (data, cb) {
+  if (!stream.write(data)) {
+    stream.once('drain', cb)
+  } else {
+    process.nextTick(cb)
+  }
+}
+```
```

Comment:
So what you’re suggesting is to actually write code as if you should *always* wait for drain. This function is just a utility to emulate a drain happening even when one won’t happen by providing an asynchronous alternative to `write()` which completes when `stream` is no longer trying to exert backpressure. I think a short explanation of how this solution only works if you serialize your `write()` calls to wait for the `cb` to be called would be helpful to clarify how this is meant to help. E.g., `to respect backpressure is to use the following. You must always wait for `cb` to be called before calling this `write()` wrapper again to respect backpressure.` Hmm, maybe that just makes it noisier.

This suggestion does simplify the code needed to respect `'drain'` by moving the part dealing with `write()`’s return value behind an “always wait for target to be ready” API. But it would still, depending on where the `write()` calls are coming from, require significant changes in the code that was originally calling `write()` to use correctly. But if the code you start with is just a series of calls to `write()` this would be straightforward enough and is rather nifty.